Revised our public CesiumJS & webpack guide

srothst1 opened the pull request:
The [current CesiumJS and Webpack guide](https://cesium.com/learn/cesiumjs-learn/cesiumjs-webpack/) is outdated. Even if a user follows the instructions exactly, they will end up with numerous errors and broken code. This branch is a revised version of the original guide. See `TUTORIAL.md` for the guide. The final step where CesiumJS is implemented may require some more time.

Feedback of all kinds would be appreciated! I am still somewhat new to webpack, so suggestions regarding webpack are particularly welcomed.

@shehzan10

srothst1 committed:
revised CesiumJS & webpack guide

liuqun commented on webpack.config.js:
```diff
@@ -1,61 +1,64 @@
         }),
         // Copy Cesium Assets, Widgets, and Workers to a static directory
-        new CopyWebpackPlugin({
+        new CopywebpackPlugin({ 
```
The trailing white space(s) should be trimmed

liuqun commented on webpack.config.js:
```diff
@@ -1,61 +1,64 @@
+    mode: 'development',
+    devtool: 'eval',
+};
+
```
The trailing blank line(s) should be trimmed too...

liuqun reviewed (commented):
Is it assumed that only the newest version of `Cesium` and `Webpack` would work together?

srothst1 committed:
updated webpack.config.js

srothst1 commented:
@liuqun Thank you for the feedback. I believe my most recent commit resolves the issues that you brought up. Does the final section of `TUTORIAL.md` covering CesiumJS look okay to you? 

We should encourage users to use the newest version of both CesiumJS and webpack.

liuqun commented on package.json:
```diff
@@ -1,38 +1,24 @@
+    "start": "node_modules/.bin/webpack serve --config webpack.config.js --open"
   },
+  "author": "",
+  "license": "ISC",
```
We should keep the **"author"** and **"license"** as *"Cesium GS, Inc."* and *"Apache-2.0"*.

so does the **"name"**, **"description"**, **"keywords"**, and **"homepage"**

liuqun commented on package.json:
```diff
@@ -1,38 +1,24 @@
-    "copy-webpack-plugin": "^6.0.3",
-    "css-loader": "^3.2.0",
-    "html-webpack-plugin": "^3.2.0",
+    "cesium": "^1.84.0",
```
I'm not quite sure whether cesium should be put into "dependencies" or "devDependencies"?

liuqun commented on src/index.html:
```diff
@@ -1,11 +1,13 @@
+  <meta charset="utf-8">
+  <!-- Include the CesiumJS JavaScript and CSS files -->
+  <script src="https://cesium.com/downloads/cesiumjs/releases/1.84/Build/Cesium/Cesium.js"></script>
+  <link href="https://cesium.com/downloads/cesiumjs/releases/1.84/Build/Cesium/Widgets/widgets.css" rel="stylesheet">
```
CDN is ok for most web applications, but painful for intranet environment with no internet connections...

> I'm not quite sure whether cesium should be put into "dependencies" or "devDependencies"

srothst1 replied:
The purpose of this guide is to provide a general framework for users to get up and running with CesiumJS and Webpack. In the future, we can create a guide that caters better to applications with no internet access.

ggetz replied:
This tutorial already installs a local copy of CesiumJS via npm. Unless I'm missing something, there should be no need to grab this from a CDN, we can instead use the version of both of these files in node modules.

liuqun commented on webpack.config.js:
```diff
@@ -1,61 +1,63 @@
     },
     output: {
-        filename: '[name].js',
-        path: path.resolve(__dirname, 'dist')
+        filename: 'app.js',
```
`filename: '[name].js',` is better when there are multipule entries

liuqun commented on webpack.config.js:
```diff
@@ -1,61 +1,63 @@
-                { from: 'node_modules/cesium/Build/Cesium/Assets', to: 'Assets' },
-                { from: 'node_modules/cesium/Build/Cesium/Widgets', to: 'Widgets' }
-            ],
+                { from: path.join(cesiumSource, cesiumWorkers), to: 'Workers' },
+                { from: path.join(cesiumSource, 'Assets'), to: 'Assets' },
+                { from: path.join(cesiumSource, 'Widgets'), to: 'Widgets' }
```
```
const CESIUM_HOME_DIR = path.resolve(__dirname, 'node_modules/cesium')
const CESIUM_SOURCE_DIR = path.resolve(CESIUM_HOME_DIR, 'Source')
const CESIUM_BUILD_DIR = path.resolve(CESIUM_HOME_DIR, 'Build')

let cesiumWorkersDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Workers')
let cesiumThirdPartyDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/ThirdParty')
let cesiumAssetsDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Assets')
let cesiumWidgetsDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Widgets')

// ...
                { from: cesiumWorkersDir, to: 'Workers' },
                { from: cesiumThirdPartyDir, to: 'ThirdParty' }, 
                { from: cesiumAssetsDir, to: 'Assets' },
                { from: cesiumWidgetsDir, to: 'Widgets' }
```

liuqun reviewed (commented):
I like this TUTORIAL.md.
And, maybe we should add a URL link for the readers to view the newest TUTORIAL.md on Github?

liuqun commented on TUTORIAL.md:
```diff
@@ -0,0 +1,445 @@
+        // Copy Cesium Assets, Widgets, and Workers to a static directory
+        new CopywebpackPlugin({ 
+            patterns: [
+                { from: path.join(cesiumSource, cesiumWorkers), to: 'Workers' },
+                { from: path.join(cesiumSource, 'Assets'), to: 'Assets' },
+                { from: path.join(cesiumSource, 'Widgets'), to: 'Widgets' }
```
```
const CESIUM_HOME_DIR = path.resolve(__dirname, 'node_modules/cesium')
const CESIUM_SOURCE_DIR = path.resolve(CESIUM_HOME_DIR, 'Source')
const CESIUM_BUILD_DIR = path.resolve(CESIUM_HOME_DIR, 'Build')

let cesiumWorkersDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Workers')
let cesiumThirdPartyDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/ThirdParty')
let cesiumAssetsDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Assets')
let cesiumWidgetsDir = path.resolve(CESIUM_BUILD_DIR, 'Cesium/Widgets')

// ...
                { from: cesiumWorkersDir, to: 'Workers' },
                { from: cesiumThirdPartyDir, to: 'ThirdParty' }, 
                { from: cesiumAssetsDir, to: 'Assets' },
                { from: cesiumWidgetsDir, to: 'Widgets' }
```

I like this `TUTORIAL.md`.
And, maybe we should add a URL link for the readers to view the newest TUTORIAL.md on Github?

srothst1 committed:
minor updates

srothst1 commented:
@liuqun 

> I like this TUTORIAL.md.
> And, maybe we should add a URL link for the readers to view the newest TUTORIAL.md on Github?

Thank you for the feedback. I incorporated the changes that I feel are appropriate for this tutorial. Given the nature of this tutorial, the breadth and depth of the information provided must not overwhelm a new user. Thus, features like offline functionality are not necessary here.

liuqun reviewed (approved):
Shall we set up a collection for more unofficial cesium tutorials? 
Like:
- https://github.com/topics/awesome
- https://github.com/topics/cesium <br> https://github.com/topics/cesium-demo

srothst1 commented:
@liuqun sure thing!

liuqun commented:
- [ ] https://github.com/martinageradams/vue-cesium-webpack-starter
- [ ] ...

srothst1 committed:
updated webpack-dev-server

srothst1 commented:
While what we have now is probably sufficient for getting community members up and running with CesiumJS and webpack, we should include information on how to import ES6 modules directly. This information could go in the **Advanced webpack configurations and Resources** section.

mramato commented on package.json:
```diff
@@ -1,38 +1,24 @@
+    "build": "node_modules/.bin/webpack --config webpack.config.js",
+    "start": "node_modules/.bin/webpack serve --config webpack.config.js --open"
   },
+  "author": "Sam Rothstein",
```
Make sure author/keyword/license information is all reverted back to the previous information.  Also update the urls so they point to cesium.com

mramato replied:
Not sure why this comment was resolved, but please don't forget about the above.

ggetz commented on src/index.js:
```diff
@@ -1,18 +1,19 @@
-import { Cesium3DTileset, createWorldTerrain, IonResource, Viewer } from 'cesium';
-import "cesium/Build/Cesium/Widgets/widgets.css";
-import "./css/main.css";
+Cesium.Ion.defaultAccessToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiJlYWE1OWUxNy1mMWZiLTQzYjYtYTQ0OS1kMWFjYmFkNjc5YzciLCJpZCI6NTc3MzMsImlhdCI6MTYyNzg0NTE4Mn0.XcKpgANiY19MC4bdFUXMVEBToBmqS8kuYpUlxJHYZxk';
```
Again, unless I'm missing something, this entire file is using the global `Cesium` object which is being included via the CDN in `index.html` and circumvents webpack entirely.

We'll need to import modules from the locally installed `cesium` npm package in order to have webpack actually involve Cesium in the bundle. If we want to keep the `Cesium` object to be more consistent with Sandcastle code, we should be able to do something like `import * as Cesium from "cesium"`. 

It may be worth noting that this includes all of Cesium and the overall bundle size will be larger, even if the app is not using all of the modules. In practice, it shouldn't be _too_ big of a difference unless the app is not actually rendering things. (Although it looks like you are on top of this based on this comment https://github.com/CesiumGS/cesium-webpack-example/pull/28#issuecomment-905077573)

srothst1 replied:
I just updated both `index.html` and `index.js`. My changes should address the concerns that you brought up. In `index.js`, I believe that we are now importing modules from the locally installed `cesium` npm package. In `index.html`, I am no longer including the `Cesium` object via CDN. I am still relatively new to webpack, so feedback of all kinds is welcomed!

ggetz reviewed (commented):
Thanks @srothst1! Great to see this tutorial getting some attention and happy that we're using the up to date versions of everything here.

I have one concern with how we are importing CesiumJS. I think we should get that cleared up before diving into any of the particulars.

srothst1 committed:
updated index.js

srothst1 commented:
@ggetz thank you for taking the time to look over this pull request. Once completed, this will be a terrific resource for the CesiumJS community! I agree completely, let's ensure that we are importing CesiumJS correctly and then do a deep dive into the actual tutorial.

ggetz commented on src/index.html:
```diff
@@ -1,11 +1,10 @@
-  <script src="https://cesium.com/downloads/cesiumjs/releases/1.84/Build/Cesium/Cesium.js"></script>
-  <link href="https://cesium.com/downloads/cesiumjs/releases/1.84/Build/Cesium/Widgets/widgets.css" rel="stylesheet">
-</head>
+  <script src="../node_modules/cesium/Build/Cesium/Cesium.js"></script>
```
I think you can omit this tag. Since you are no longer using the global cesium object, this is no longer needed.

srothst1 replied:
I just removed this tag.

ggetz commented on src/index.html:
```diff
@@ -1,11 +1,12 @@
+<head>Cesium and Webpack</head>
+  <meta charset="utf-8">
+  <script src="../node_modules/cesium/Build/Cesium/Cesium.js"></script>
+  <link href="../node_modules/cesium/Build/Cesium/Widgets/widgets.css" rel="stylesheet">
```
If you include the cesium css files this way, I don't believe there's any reason to have the css loader. Serving up the static files via the dev server and copied with the CopyWebpackPlugin should be enough.

ggetz commented on src/index.js:
```diff
@@ -1,18 +1,21 @@
+import * as Cesium from "../node_modules/cesium"
```
The webpack file has aliased the Cesium Source folder as `cesium`, so this could be 

```suggestion
import * as Cesium from "cesium"
```

It's pretty common to use aliases, or at least have an alias for `node_modules` to avoid relative paths.

If you don't want to use the source files directly, you can configure the alias to point at the top level cesium path as you use here.

If you chose not to use the alias, it should be removed from the config file altogether.

ggetz reviewed (commented):
Awesome, thanks @srothst1 ! I have a few more comments. Additionally you might want to check the CI logs and see what is causing it to error.

srothst1 committed:
updating Cesium alias

Co-authored-by: Gabby Getz <gabby@cesium.com>

srothst1 committed:
removed tag

srothst1 committed:
removed whitespace

srothst1 committed:
updated dependencies

srothst1 commented:
@ggetz thank you for adding some additional comments! In terms of the Travis CI build, I am currently getting the following error.

`npm ERR! missing script: release`

I am unsure why the build is running the line `$ npm run release`. Do you have any suggestions? I am going to continue looking into the issue later this afternoon.

ggetz commented:
@srothst1 Travis is configured in [.travis.yml](https://github.com/CesiumGS/cesium-webpack-example/blob/main/.travis.yml).

It looks like you changed the names of the scripts in https://github.com/CesiumGS/cesium-webpack-example/pull/28/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R6. You'll need to update that file to use the new script names.

srothst1 committed:
updated .travis.yml

mramato commented on src/index.js:
```diff
@@ -1,18 +1,23 @@
+import * as Cesium from "../node_modules/cesium"
```
`import *` is an anti-pattern we should not use in our own code.  Only import the parts of the Cesium API that you need.  This helps with tree-shaking so that only the parts of Cesium you actually used get imported.

srothst1 replied:
I just updated `index.js` to only include the imports that are needed for this application.

srothst1 commented:
@ggetz thanks for the clarification - I just updated `.travis.yml`. I believe that `npm run build` and `npm run start` are the appropriate scripts to run.

srothst1 committed:
updated imports

srothst1 committed:
updated index.js and index.html

srothst1 committed:
1.0.0-canary.0

srothst1 committed:
trying to resolve Travis CI issues

srothst1 committed:
trying to resolve Travis CI issues part 2

srothst1 committed:
trying to resolve Travis CI issues part 3

srothst1 committed:
trying to resolve Travis CI issues part 4

srothst1 committed:
trying to resolve Travis CI issues part 5

srothst1 committed:
trying to resolve Travis CI issues part 6

srothst1 committed:
trying to resolve Travis CI issues part 7

srothst1 committed:
trying to resolve Travis CI issues part 8

srothst1 committed:
trying to resolve Travis CI issues part 9

srothst1 committed:
trying to resolve Travis CI issues part 10

srothst1 commented:
My immediate goal is to make sure that the code is working properly. Once this is complete, I will continue editing the guide. I believe that I have incorporated all of the feedback related to `index.js` and `index.html`. 

I am currently struggling to resolve a Travis CI-related issue. The command `npm run release` is giving me the following error.
```
[webpack-cli] Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
```
I added the following lines to `package.json`
```
    "release": "node_modules/.bin/webpack --config webpack.release.config.js",
    "serve-release": "node_modules/.bin/webpack-dev-server --config webpack.release.config.js --content-base /dist"
```
and created the file `webpack.release.config.js`. I have also tried strategies suggested in the following threads but have had no luck.
- https://www.codegrepper.com/code-examples/whatever/Webpack+has+been+initialized+using+a+configuration+object+that+does+not+match+the+API+schema%24
- https://stackoverflow.com/questions/42060243/invalid-configuration-object-webpack-has-been-initialised-using-a-configuration
- https://stackoverflow.com/questions/43002099/rules-vs-loaders-in-webpack-whats-the-difference/43805263#43805263

I am relatively new to Travis CI, which makes this issue particularly challenging for me.

ggetz commented:
@srothst1 Do you get the same errors when running `npm run release` on your machine? Or is this only happening in Travis?

srothst1 commented:
@ggetz I get the same errors when I run `npm run release` on my machine.

ggetz commented:
OK, so that means the problem is not with Travis. If you get that script running on your machine, it should work similarly in Travis.

Do you plan on having a release configuration as part of this tutorial? A release configuration omits any debugging tools and may have some additional optimization as compared to the dev config. If you do not plan on including it, you can omit the `release` script from both `package.json` and `travis.yml` entirely.

srothst1 committed:
removed the release configuration

srothst1 committed:
removed release.config

srothst1 commented:
@ggetz thank you for the additional information. For now, including a release configuration does not seem necessary. Are there any additional edits that should be made to the code part of this tutorial? If not, I will resume editing `TUTORIAL.md`.

ggetz commented on src/index.js:
```diff
@@ -1,18 +1,23 @@
+import {Ion, Viewer, createWorldTerrain, createOsmBuildings, Cartesian3, Math } from "../node_modules/cesium"
```
```suggestion
import { Ion, Viewer, createWorldTerrain, createOsmBuildings, Cartesian3, Math } from "../node_modules/cesium"
```

ggetz replied:
Did you have any luck getting this to work with the cesium alias, ie `from "cesium"` ?

ggetz reviewed (commented):
I think that's all I have for the code @srothst1. Thanks!

srothst1 committed:
adding a whitespace to src/index.js

Co-authored-by: Gabby Getz <gabby@cesium.com>

srothst1 committed:
updating tutorial

srothst1 committed:
updating tutorial pt 2

srothst1 committed:
updated heading size

ebogo1 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,433 @@
+    }
+};
+```
+6. To defined `index.html` and inject our bundle into that page we will be using a webpack [plugin](https://webpack.js.org/concepts/#plugins) called [html-webpack-plugin](https://webpack.js.org/concepts/#plugins). Use the command `npm install --save-dev html-webpack-plugin` to install the necessary plugin.
````
```suggestion
6. To define `index.html` and inject our bundle into that page we will be using a webpack [plugin](https://webpack.js.org/concepts/#plugins) called [html-webpack-plugin](https://webpack.js.org/concepts/#plugins). Use the command `npm install --save-dev html-webpack-plugin` to install the necessary plugin.
```

ebogo1 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,433 @@
+```
+3. Tell the development server to serve files in the `dist` folder. This can be accomplished by adding the following code to `webpack.config.js`
+```
+// development server options
````
Is this `devServer` object supposed to be in the next block of code? I'm not clear on where this is supposed to go.

srothst1 replied:
Good point. I had this same issue when I went through the tutorial the first time. To be honest, I am not sure if the order matters. Should I suggest that users put this code at the bottom of `module.exports` in `webpack.config.js`?

ebogo1 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,433 @@
+// The path to the CesiumJS source code
+const cesiumSource = 'node_modules/cesium/Source';
+const cesiumWorkers = '../Build/Cesium/Workers';
+const CopywebpackPlugin = require('copy-webpack-plugin');
```
The `CopywebpackPlugin` isn't talked about until later in the tutorial. Also, should it be `CopyWebpackPlugin` for consistency?

srothst1 replied:
Good catch - just added a commit that addresses this.

ebogo1 replied:
Looks like it's still `CopywebpackPlugin` - the same capitalization is also repeated in later sections

ebogo1 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,433 @@
+        })
+    ],
+    // development server options
+    devServer: {
```
If I add the `devServer` property with the `contentBase` key, I get this error in console after running `npm start`:

```
[webpack-cli] Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'contentBase'. These properties are valid:
object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, setupExitSignals?, static?, watchFiles?, webSocketServer? }
 ```

srothst1 replied:
@ebogo1 thank you for bringing this to my attention! I found various stackoverflow threads that address similar issues. 

- https://stackoverflow.com/questions/62967698/if-i-set-contentbase-path-of-dev-server-webpack-doesnt-serve-updated-files

- https://stackoverflow.com/questions/67926476/webpack-dev-server-config-contentbase-not-working-in-latest-version

Upon further thought, I believe that `devServer` can be omitted from `webpack.config.js`. Based on our previous code, the files should automatically be added to the `dist` folder. Can you please confirm with me that this works on your end?

srothst1 replied:
I revised `TUTORIAL.md` to reflect these updates.

srothst1 committed:
Updated step 6 of TUTORIAL.md

Co-authored-by: Eli Bogomolny <31491650+ebogo1@users.noreply.github.com>

srothst1 committed:
CopywebpackPlugin -> CopyWebpackPlugin

srothst1 committed:
updated heading size

srothst1 committed:
removed devServer

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+1. Create a new directory for your app.
+2. Open a console window and navigate to the directory
+3. Run `npm init` and enter in the requested details about your application. If you are unsure about any prompts, press `Enter` to use the default value. You can modify your selections at any point in the `package.json` file that was created.
```
```suggestion
3. Run `npm init` and enter in the requested details about your application. If you are unsure about any prompts, press `Enter` to use the default value. You can modify these details at any point in the `package.json` file that `npm init` creates.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+## Initialize an app with npm
+
+1. Create a new directory for your app.
+2. Open a console window and navigate to the directory
```
```suggestion
2. Open a console window and navigate to the directory.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+# CesiumJS and Webpack
+
+[Webpack](https://webpack.js.org/) is a popular and powerful tool for bundling JavaScript modules. It allows developers to structure their code and assets in an intuitive way and to load different kinds of files as needed with simple require statements. When building, it will trace code dependencies and pack these modules into one or more bundles that are loaded by the web browser.
```
```suggestion
[Webpack](https://webpack.js.org/) is a popular and powerful tool for bundling JavaScript modules. It allows developers to structure their code and assets in an intuitive way and to load different kinds of files as needed with simple `require` statements. When building, it will trace code dependencies and pack these modules into one or more bundles that are loaded by the web browser.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+This will serve as our boilerplate JavaScript code.
+
+## Install and configure webpack
+1. Install webpack by running `npm install --save-dev webpack`. The files `package-lock.json` and `package.json` should appear in your main directory. You should also see a new folder in your main directory named `node_modules`.
```
Should you include a step or comment here about what to do in case these files and folders don't appear? If it's a relatively easy mistake to make, I think you should, but if it's very unlikely, then you don't need to.

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+    }
+};
+```
+In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundel `app.js` to the folder `dist`, that webpack will create at runtime.
````
```suggestion
In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundle `app.js` to the `dist` folder, which webpack will create at runtime.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+};
+```
+In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundel `app.js` to the folder `dist`, that webpack will create at runtime.
+4. Webpack loads everything like a module. [loaders](https://webpack.js.org/concepts/#loaders) are used to load CSS and other asset files. Install the [style-loader](https://webpack.js.org/loaders/style-loader/#src/components/Sidebar/Sidebar.jsx), [css-loader](https://webpack.js.org/loaders/css-loader/), and [url-loader](https://webpack.js.org/loaders/url-loader/) using `npm install --save-dev st ...
````
> Webpack loads everything like a module.

This line doesn't make a ton of sense to me. Here's a suggestion for rewording based on my understanding of the webpack documentation. 

```suggestion
4. In order to load CSS and other asset files with webpack, you'll need to install some [loaders](https://webpack.js.org/concepts/#loaders). Install the [style-loader](https://webpack.js.org/loaders/style-loader/#src/components/Sidebar/Sidebar.jsx), [css-loader](https://webpack.js.org/loaders/css-loader/), and [url-loader](https://webpack.js.org/loaders/url-loader/) using `npm install --save-dev style-loader css-loader url-loader`. Feel free to install any other loaders you may need in the future. Loaders can be installed at any point during this process.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+```
+In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundel `app.js` to the folder `dist`, that webpack will create at runtime.
+4. Webpack loads everything like a module. [loaders](https://webpack.js.org/concepts/#loaders) are used to load CSS and other asset files. Install the [style-loader](https://webpack.js.org/loaders/style-loader/#src/components/Sidebar/Sidebar.jsx), [css-loader](https://webpack.js.org/loaders/css-loader/), and [url-loader](https://webpack.js.org/loaders/url-loader/) using `npm install --save-dev st ...
+5. Update `webpack.config.js` by adding two `module.rules`. The first rule should support CSS files and the second rule should support other static files. For each rule, define `test` for the types of files to load and `use` to specify the list of loaders. `webpack.config.js` should look something like
````
```suggestion
5. Update `webpack.config.js` by adding two `module.rules`. The first rule will support CSS files and the second rule will support other static files. For each rule, define `test` for the types of files to load and `use` to specify the list of loaders. Your `webpack.config.js` should resemble the following:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+    }
+};
+```
+6. To define `index.html` and inject our bundle into that page we will be using a webpack [plugin](https://webpack.js.org/concepts/#plugins) called [html-webpack-plugin](https://webpack.js.org/concepts/#plugins). Use the command `npm install --save-dev html-webpack-plugin` to install the necessary plugin.
````
```suggestion
6. To define `index.html` and inject our bundle into that page, we will be using a webpack [plugin](https://webpack.js.org/concepts/#plugins) called [html-webpack-plugin](https://webpack.js.org/concepts/#plugins). Use the command `npm install --save-dev html-webpack-plugin` to install this plugin.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+};
+```
+6. To define `index.html` and inject our bundle into that page we will be using a webpack [plugin](https://webpack.js.org/concepts/#plugins) called [html-webpack-plugin](https://webpack.js.org/concepts/#plugins). Use the command `npm install --save-dev html-webpack-plugin` to install the necessary plugin.
+7. Require html-webpack-plugin in `webpack.config.js` by adding it to the `plugins` section. Next, pass `src/index.html` as our `template`. Finally, specify the mode option for webpack by adding `mode: 'development'` to `webpack.config.js`.`webpack.config.js` should now look something like
````
The "Your"s  that I'm adding aren't really necessary, but I think they help with readability.

```suggestion
7. Require html-webpack-plugin in `webpack.config.js` by adding it to the `plugins` section. Next, pass `src/index.html` as our `template`. Finally, specify the mode option for webpack by adding `mode: 'development'` to `webpack.config.js`. Your `webpack.config.js` should now resemble the following:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+  }
+```
+
+If done correctly, `package.json` should look something like this
````
```suggestion
If done correctly, `package.json` should now resemble the code below. Note that this json file will vary based on your selections in step 3 of the **Initialize an app with npm** section.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+}
+```
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
````
```suggestion

```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+```
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
+2. Run the command `npm run build`. Install CLI for webpack if necessary. 
````
Are there any steps or commands associated with "Install CLI for webpack"?

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
+2. Run the command `npm run build`. Install CLI for webpack if necessary. 
+3. Ensure that there are no errors and your output looks something like
```
```suggestion
3. Ensure that there are no errors and your output looks something like:
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+## Run the development server
+1. We will be using a [webpack-dev-server](https://webpack.js.org/configuration/dev-server/) to serve a development build and see our application in action. Run `npm install --save-dev webpack-dev-server`.
+2. Add a `start` script to `package.json`. This script should run the development server. Be sure to set the config file via the `--config` flag and use the `--open` flag to open the application in a a browser upon execution of the command. `package.json` should look something like
```
```suggestion
2. Add a `start` script to `package.json`. This script will run the development server. Be sure to set the config file via the `--config` flag, and use the `--open` flag to open the application in a a browser upon execution of the command. `package.json` should look something like:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+  }
+}
+```
+3. Run `npm start` and open http://localhost:8080/ in a web browser.
````
```suggestion
3. Run `npm start` and open [http://localhost:8080/](http://localhost:8080/) in a web browser.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+}
+```
+3. Run `npm start` and open http://localhost:8080/ in a web browser.
+4. Ensure that you see Hello World!
````
```suggestion
4. You should now see the text "Hello World!" on the page.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+## Add CesiumJS to a webpack app
+
+CesiumJS is a large and complex library. In additional to JavaScript modules, it also includes static assets such as CSS, image, and json files. It includes web worker files to perform intensive calculations in separate threads. Unlike traditional npm modules, CesiumJS does not define an entry point because of the diverse ways in which the library is used. We’ll need to configure some additional  ...
```
```suggestion
CesiumJS is a large and complex library. In additional to JavaScript modules, it includes static assets such as CSS, image, and json files. It also includes web worker files to perform intensive calculations in separate threads. Unlike traditional npm modules, CesiumJS does not define an entry point because of the diverse ways in which the library is used. We’ll need to configure some additional options to use it with webpack.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+CesiumJS is a large and complex library. In additional to JavaScript modules, it also includes static assets such as CSS, image, and json files. It includes web worker files to perform intensive calculations in separate threads. Unlike traditional npm modules, CesiumJS does not define an entry point because of the diverse ways in which the library is used. We’ll need to configure some additional  ...
+
+First, define where CesiumJS is. This tutorial uses the source code, so webpack can include individual models and trace the dependencies. Alternatively, you can use the built (minified or unminified) version of CesiumJS. However, the modules are already combined and optimized, which gives us less flexibility.
```
```suggestion
First, define where CesiumJS is. This tutorial uses the source code, so webpack can include individual models and trace the dependencies. Alternatively, you can use the built (minified or unminified) version of CesiumJS. However, the modules are already combined and optimized in the built version, which gives us less flexibility.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+First, define where CesiumJS is. This tutorial uses the source code, so webpack can include individual models and trace the dependencies. Alternatively, you can use the built (minified or unminified) version of CesiumJS. However, the modules are already combined and optimized, which gives us less flexibility.
+
+## Install CesiumJS
+1. Install the [Cesium](https://www.npmjs.com/package/cesium) module from npm using the command `npm install --save-dev cesium`. Install the [CopyWebpackPluigin](https://webpack.js.org/plugins/copy-webpack-plugin/) module from npm using the command `npm install copy-webpack-plugin --save-dev`.
```
```suggestion
1. Install the [Cesium](https://www.npmjs.com/package/cesium) module from npm using the command `npm install --save-dev cesium`. Install the [CopyWebpackPlugin](https://webpack.js.org/plugins/copy-webpack-plugin/) module from npm using the command `npm install copy-webpack-plugin --save-dev`.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+## Install CesiumJS
+1. Install the [Cesium](https://www.npmjs.com/package/cesium) module from npm using the command `npm install --save-dev cesium`. Install the [CopyWebpackPluigin](https://webpack.js.org/plugins/copy-webpack-plugin/) module from npm using the command `npm install copy-webpack-plugin --save-dev`.
+2. Update `sourcePrefix` to tell CesiumJS that the version of AMD webpack uses to evaluate `require` statements is not compliant with the standard `toUrl` function. In addition, add a `cesium` alias so we can reference it in our app code. After adding these changes, `webpack.config.js` should look like
```
```suggestion
2. Update `sourcePrefix` to tell CesiumJS that the version of AMD that webpack uses to evaluate `require` statements is not compliant with the standard `toUrl` function. Then, add a `cesium` alias so we can reference it in our app code. After adding these changes, `webpack.config.js` should look like:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+```
+
+## Manage CesiumJS static files
+1. Now, we must make sure the static CesiumJS asset, widget, and web worker files are served and loaded correctly. Use `copy-webpack-plugin` to copy static files to the `dist` directory as part of the build process. To do this run the command `npm install --save-dev copy-webpack-plugin` and updated the plugins array in `webpack.config.js`. `webpack.config.js` should now look like
````
```suggestion
1. Now, we must make sure the static CesiumJS asset, widget, and web worker files are served and loaded correctly. You will use `copy-webpack-plugin` to copy static files to the `dist` directory as part of the build process. To do this, run the command `npm install --save-dev copy-webpack-plugin` and update the plugins array in `webpack.config.js`. Your `webpack.config.js` should now look like:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+```
+
+## Incorporate CesiumJS
+1. Updated `index.js` with CesiumJS starter code:
````
```suggestion
1. Update `index.js` with CesiumJS starter code:
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+});
+```
+This code initializes the Cesium Viewer, adds Cesium OSM Buildings to the terrain, and moves the `Camera` to San Francisco. 
+2. Update `index.html` with the CesiumContainer
````
```suggestion
2. Update `index.html` with the `CesiumContainer`.
```

argallegos commented on TUTORIAL.md:
````diff
@@ -0,0 +1,386 @@
+</html>
+```
+3. Run the command `npm run build`. Ensure that the project is built correctly.
+4. Run `npm start` and open http://localhost:8080/ in a web browser to see the CesiumJS viewer.
````
```suggestion
4. Run `npm start` and open [http://localhost:8080/](http://localhost:8080/) in a web browser to see the CesiumJS viewer.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+![South San Francisco](./screenshots/ssf2.png)
+
+Copy and paste your favorite Sandcastle example. For example, [The Particle System Fireworks](https://sandcastle.cesium.com/?src=Particle%20System%20Fireworks.html) demo makes for a fun demo.
```
```suggestion
Copy and paste your favorite Sandcastle example. Need inspiration? Try the [Particle System Fireworks](https://sandcastle.cesium.com/?src=Particle%20System%20Fireworks.html) demo.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+## Additional Resources
+
+The official cesium-webpack-example  repo contains the minimal webpack configuration, the hello world code covered in this tutorial, and instructions for optional code configurations.
```
```suggestion
The official [cesium-webpack-example repository](https://github.com/CesiumGS/cesium-webpack-example) contains the minimal webpack configuration, the hello world code covered in this tutorial, and instructions for optional code configurations.
```

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+The official cesium-webpack-example  repo contains the minimal webpack configuration, the hello world code covered in this tutorial, and instructions for optional code configurations.
+
+For a tour of CesiumJS feature to include in your new app, see the Cesium Workshop Tutorial.
```
I don't think the Cesium Workshop Tutorial still exists. If you can find it, link it here, and if not, delete this line.

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+For a tour of CesiumJS feature to include in your new app, see the Cesium Workshop Tutorial.
+
+Explore demos in Sandcastle and see the CesiumJS Documentation.
```
Feel free to change these links if there's a better page.
```suggestion
Learn CesiumJS with the [CesiumJS Documentation](https://cesium.com/learn/cesiumjs-learn/) and explore [Sandcastle](https://sandcastle.cesium.com/) demos to see CesiumJS in action.
```

srothst1 replied:
Will likely link to the CesiumJS tutorials - [CesiumJS tutorials](https://cesium.com/learn/cesiumjs-learn/)

argallegos commented on TUTORIAL.md:
```diff
@@ -0,0 +1,386 @@
+
+Explore demos in Sandcastle and see the CesiumJS Documentation.
+
+To learn more about webpack, take a look at webpack Concepts, or dive into the Documentation.
```
```suggestion
Want to learn more about webpack? Take a look at webpack [Concepts](https://webpack.js.org/concepts/), or dive into the [webpack documentation](https://webpack.js.org/configuration/).
```

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
Update TUTORIAL.md

Co-authored-by: Alex Gallegos <39537389+argallegos@users.noreply.github.com>

srothst1 committed:
updated webpack links

ebogo1 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,384 @@
+
+## Manage CesiumJS static files
+1. Now, we must make sure the static CesiumJS asset, widget, and web worker files are served and loaded correctly. Use `copy-webpack-plugin` to copy static files to the `dist` directory as part of the build process. To do this run the command `npm install --save-dev copy-webpack-plugin` and updated the plugins array in `webpack.config.js`. `webpack.config.js` should now look like
+```
````
I noticed that `copy-webpack-plugin` isn't actually used until this step even though it's installed and referenced in the previous one. Could you move those lines to this step instead?

srothst1 replied:
Which lines should be moved?

ebogo1 replied:
In step 1 of "Install CesiumJS": 

> Install the CopyWebpackPluigin module from npm using the command `npm install copy-webpack-plugin --save-dev`

Looks like this line is just repeated in this step (where it should be) so the line in Install CesiumJS can just be removed unless I'm missing something

srothst1 replied:
Ahhh I see - I just removed the extra line.

srothst1 committed:
updated TUTORIAL.md with more details

ebogo1 commented:
On a nitpicky side note, I'd force push the final changes to TUTORIAL.md to squash the 26+ commits with the same or nearly identical messages..

srothst1 committed:
updated CopyWebpackPlugin

srothst1 committed:
removed extra line

srothst1 commented:
@argallegos thank you very much for the detailed feedback on `TUTORIAL.md`. I implemented most of the changes that you suggested :rocket: 

@ebogo1 @ggetz can you please confirm that everything looks correct before I merge?

srothst1 committed:
updated author/keyword/licesne information

srothst1 committed:
updated package.json

shehzan10 commented:
> can you please confirm that everything looks correct before I merge?

@srothst1 Please do not self-merge this. Please make sure that one of the PR reviewers are merging once all the requirements are met.

shehzan10 commented on package.json:
```diff
@@ -1,38 +1,61 @@
+    "name": "Cesium GS, Inc.",
+    "url": "https://cesium.com"
+  },
+  "contributors": [
```
@srothst1 We don't need this section in package.json.

srothst1 replied:
@shehzan10 sounds good - I will remove the "contributors" section.

srothst1 committed:
removed the contributors section

ebogo1 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,383 @@
+In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundel `app.js` to the folder `dist`, that webpack will create at runtime.
+4. Webpack loads everything like a module. [loaders](https://webpack.js.org/concepts/#loaders) are used to load CSS and other asset files. Install the [style-loader](https://webpack.js.org/loaders/style-loader/#src/components/Sidebar/Sidebar.jsx), [css-loader](https://webpack.js.org/loaders/css-loader/), and [url-loader](https://webpack.js.org/loaders/url-loader/) using `npm install --save-dev st ...
+5. Update `webpack.config.js` by adding two `module.rules`. The first rule should support CSS files and the second rule should support other static files. For each rule, define `test` for the types of files to load and `use` to specify the list of loaders. `webpack.config.js` should look something like
+```
````
Not sure what the fix is but I noticed the numbering loses continuity here: 

![image](https://user-images.githubusercontent.com/31491650/136324223-01477f6a-17dd-4a3e-81f2-8cc9f46b3ae1.png)

srothst1 replied:
Thanks for catching this - I just updated the Markdown file.

srothst1 committed:
Update TUTORIAL.md

boeckMt commented on package.json:
```diff
@@ -1,38 +1,43 @@
-    "start": "node_modules/.bin/webpack-dev-server --config webpack.config.js",
-    "release": "node_modules/.bin/webpack --config webpack.release.config.js",
-    "serve-release": "node_modules/.bin/webpack-dev-server --config webpack.release.config.js --content-base /dist"
+  "dependencies": {
```
Should there not only be one dependency 

```
"cesium": "^1.84.0",
```

and all others are being `devDependencies`?

ggetz commented on package.json:
```diff
@@ -1,38 +1,43 @@
+    "url": "https://cesium.com"
+  },
   "keywords": [
-    "cesium",
```
This is nitpick-y, but `cesium` would still be a good keyword.

srothst1 replied:
No not at all! I think this is a great idea - I just added it back.

ggetz commented on package.json:
```diff
@@ -1,38 +1,43 @@
-    "release": "node_modules/.bin/webpack --config webpack.release.config.js",
-    "serve-release": "node_modules/.bin/webpack-dev-server --config webpack.release.config.js --content-base /dist"
+  "dependencies": {
+    "cesiumjs": "^1.10.0",
```
`cesiumjs` shouldn't be here at all - Use `cesium` which is already included above in "devDependencies".

srothst1 replied:
Thanks for catching this. I was a little bit confused because we are using two different names to import CesiumJS - "cesium" and "cesiumjs"
Regardless, I removed the `"cesiumjs": "^1.10.0"` dependency.

ggetz commented on webpack.config.js:
```diff
@@ -1,38 +1,38 @@
-        mainFields: ['module', 'main']
+        alias: {
+            // CesiumJS module name
+            cesium: path.resolve(__dirname, cesiumSource)
```
You've defined the `cesium` alias here, but I don't see it used in `src/index.js`.

It's pretty atypical to define relative paths to dependencies in `node_modules` from source code like what's currently happening in `index.js`. Instead, you usually rely on the package name itself which [webpack resolves from `node_modules` just like node resolves](https://webpack.js.org/configuration/resolve/#resolvemodules).

At first, I would have suggested omitting the `resolve` property altogether, which would make the first line in `index.js`
```
import { Ion, Viewer, createWorldTerrain, createOsmBuildings, Cartesian3, Math } from "cesium";
```

However, there's an issue when importing `Widgets/widgets.css`. In regards to css files with cesium and webpack, there's actually [an issue in the CesiumJS](https://github.com/CesiumGS/cesium/issues/9212) repo advising the best configuration. Basically, we don't explicitly export anything other than JS files as part of the cesium npm package, which means webpack does not allow us to resolve CSS files like js modules.

After testing this out myself, I think we should keep the above alias, but configure the Cesium's main module file so that Cesium can be resolved without a relative path. So in `webpack.config.js`, `resolve` becomes

```
    resolve: {
        alias: {
            cesium: path.resolve(__dirname, cesiumSource)
        },
        mainFiles: ['index', 'Cesium']
    },
 ```
 and the first few lines of `index.js` become
 ```
import { Ion, Viewer, createWorldTerrain, createOsmBuildings, Cartesian3, Math } from "cesium";
import "cesium/Widgets/widgets.css";
```

srothst1 replied:
I was not aware of the issue in the CesiumJS repo that you shared. It looks like @thw0rted offered to submit a PR that resolves at least some of the bugs. In the meantime, I just added a commit that implements the suggestions that you gave.

thw0rted replied:
IMHO, instead of a `resolve.alias` entry in Webpack, it would be ideal if you could set up the `package.json` so that `import { Viewer } from "cesium"` just works for everybody -- webpack, rollup, or node.  It's been a minute since I reviewed the relevant fields but I think this should be practical, at least as long as you're only supporting webpack 5+.

ggetz replied:
Thanks for the suggestion! I believe that is captured in the issue I linked to above. While that change is out of the scope of this PR and repo, we can adjust this example if/when Cesium's `package.json` file is updated.

ggetz commented on src/css/main.css:
```diff
@@ -1,7 +0,0 @@
-html, body, #cesiumContainer {
```
Removing this css leaves a white border around the cesium viewer and causes the page to have a scrollbar, which looks pretty unpolished IMO. However, I see that this is also omitted from the [quickstart guide on the website](https://cesium.com/learn/cesiumjs-learn/cesiumjs-quickstart/). If the goal of this is to conform as closely to that tutorial as possible while using webpack, that's OK.

srothst1 replied:
Nice catch. I initially assumed that the goal was to mirror our previous demos as closely as possible. Ultimately, the community will use this guide as a reference and likely replace most of the code. For that reason, I think it is probably fine to add the CSS file back since they will almost certainly need a CSS file for their project. I just added it back.

ggetz reviewed (commented):
Thanks @srothst1 ! My main concern is that https://github.com/CesiumGS/cesium-webpack-example/pull/28#discussion_r698533863 is followed up on, so I added some more detail regarding aliases below.

srothst1 committed:
updated the dependencies in package.json

srothst1 committed:
updated the dependencies in package.json

srothst1 committed:
updated the webpack resolve statement and the js imports

srothst1 committed:
added back css

srothst1 committed:
added Gabby's most recent suggestions to the tutorial

ggetz commented on webpack.config.js:
```diff
@@ -1,38 +1,38 @@
+        alias: {
+            cesium: path.resolve(__dirname, cesiumSource)
+        },
+        mainFiles: ['module', 'main', 'Cesium']
```
Sorry - I corrected my comment - I believe this needs to be `[ 'index', 'Cesium' ]`.

srothst1 replied:
I just added a commit that fixes this :+1:

ggetz commented on package.json:
```diff
@@ -1,38 +1,44 @@
+    "es6": "^0.0.7",
+    "querystring-es3": "^0.2.1",
+    "release": "^6.3.0",
+    "webpack": "^5.51.1"
```
I think all of these are `devDependencies` in this case, as suggested above.

thw0rted replied:
I don't see where the other 3 dependencies (es6, querystring, release) are actually being used.  Maybe I'm missing something?

srothst1 replied:
I read through [this thread](https://stackoverflow.com/questions/18875674/whats-the-difference-between-dependencies-devdependencies-and-peerdependencies)  for some context and added a commit with the update. We now do not have any "dependencies" - is that correct?

ggetz replied:
That's my thinking, yes. The packaged version of this app would not need any of these packages installed to run. 

And, yes, please double check if all of these packages are being used, and if not, remove them.

srothst1 replied:
@ggetz good point :smile: I just ran the command `depcheck` to find the unused npm packages in `package.json` as suggested [here](https://stackoverflow.com/questions/22675725/find-unused-npm-packages-in-package-json). It initially looked like the following dependencies were not necessary.
```
Unused devDependencies
* sass-loader
* strip-pragma-loader
* webpack-cli
* webpack-dev-server
* es6
* querystring-es3
* release
```
However, when running `npm run build` and `npm run start`, I realized that `webpack-dev-server` and `webpack-cli` are actually necessary. I just pushed a commit that addresses this :+1:

ggetz commented on src/css/main.css:
```diff
@@ -1,5 +1,5 @@
-html, body, #cesiumContainer {
-	width: 100%;
+html, body, #cesiumContainer {	
+    width: 100%;
```
This file has some inconsistent formatting with spaces and tabs. Could you clean it up?

srothst1 replied:
Yup, just removed the extra spaces after `html, body, #cesiumContainer {`

ggetz reviewed (commented):
Thanks @srothst1 ! Looking good, just some final tweaks.

srothst1 committed:
updated main files

srothst1 committed:
updated the devDependencies

srothst1 committed:
removed white space in css file

srothst1 commented:
@ggetz I tested out the project after adding all of the recent changes and it still works :+1:

srothst1 committed:
updated npm packages

ggetz commented:
Thanks @srothst1 ! This is looking good to me. Thanks for your patience with the review too. Since this is a tutorial, its important we have all of our ducks in a row.

Is there anyone else that needs an eye on this or is it ready to be merged?

shehzan10 commented on README.md:
```diff
@@ -74,5 +74,3 @@ Pull requests are appreciated. Please use the same [Contributor License Agreemen
 ---
 
 Developed by the Cesium team.
-
```
Please change all URLs in this file from `AnalyticalGraphicsInc` -> `CesiumGS`, and `cesiumjs.org` -> `cesium.com`, or use best judgement as needed.

srothst1 replied:
Nice catch. You may have noticed that the `README.MD` file in the branch was simply a relic from the last webpack tutorial that we released. It is mostly relevant to people who view our GitHub page, not users accessing our tutorial. I just added a commit that addresses this issue.

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+[Webpack](https://webpack.js.org/) is a popular and powerful tool for bundling JavaScript modules. It allows developers to structure their code and assets in an intuitive way and to load different kinds of files as needed with simple require statements. When building, it will trace code dependencies and pack these modules into one or more bundles that are loaded by the web browser.
+
+In this tutorial, we will build a simple web app from the ground up using webpack, and then cover the steps to integrate the [Cesium npm module](https://www.npmjs.com/package/cesium). This is a good place to start if you’d like to use CesiumJS to develop a web application. If you’re new to Cesium and are looking to learn to build your first sample app, take a look at our [Getting Started Tutorial ...
```
Our tutorials language always addresses the audience. For example "In this tutorial, you will build...". See our website tutorials for more.

Please take a pass through the full document for these changes.

srothst1 replied:
I just added a commit that addresses this feedback. I originally used "we" in various places because I wanted the tutorial reader to feel welcomed - almost as if I was working with them to get webpack running in their development environment.

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+[Webpack](https://webpack.js.org/) is a popular and powerful tool for bundling JavaScript modules. It allows developers to structure their code and assets in an intuitive way and to load different kinds of files as needed with simple require statements. When building, it will trace code dependencies and pack these modules into one or more bundles that are loaded by the web browser.
+
+In this tutorial, we will build a simple web app from the ground up using webpack, and then cover the steps to integrate the [Cesium npm module](https://www.npmjs.com/package/cesium). This is a good place to start if you’d like to use CesiumJS to develop a web application. If you’re new to Cesium and are looking to learn to build your first sample app, take a look at our [Getting Started Tutorial ...
```
`Cesium npm module` -> `Cesium NPM module`

shehzan10 replied:
Also please be consistent in the use of Cesium vs CesiumJS. I would recommend using CesiumJS throughout.

thw0rted replied:
Per [npm's own docs](https://docs.npmjs.com/about-packages-and-modules) they prefer lower-case "npm pacakges".  ("Cesium npm package" would be more technically correct.)

srothst1 replied:
I also took some time to look over the npm documentation. I believe that @thw0rted is correct here. "Cesium npm package" is technically the right phrase to use. @shehzan10 is there other npm documentation that you can share with me that suggests otherwise? For now, I added a commit with the phrasing "Cesium npm package". I can always update this later if necessary.

thw0rted replied:
![image](https://user-images.githubusercontent.com/943160/138303340-a2eaf140-263a-4df4-9cd7-903a6fe69682.png)

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+# CesiumJS and Webpack
+
+[Webpack](https://webpack.js.org/) is a popular and powerful tool for bundling JavaScript modules. It allows developers to structure their code and assets in an intuitive way and to load different kinds of files as needed with simple require statements. When building, it will trace code dependencies and pack these modules into one or more bundles that are loaded by the web browser.
+
```
In our website tutorials, we've built the "You'll learn how to" as a bulleted list. I would recommend that here too. The paragraph below is good, but it is good to summarize it to the user as well.

srothst1 replied:
I just took a look at the [Cesium for Unreal Quickstart](https://cesium.com/learn/unreal/unreal-quickstart/) tutorial for reference. I merged the first two paragraphs into one concise paragraph and added a "You'll learn how to" section :+1: 

Added a commit addressing this.

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+## Prerequisites
+- A basic understanding of the command line, JavaScript, and web development.
+- An IDE or code editor. Developers on the Cesium team members use Visual Studio Code, but a minimal code editor such as Sublime Text will also work.
```
Structure fix needed: "Developers on the Cesium team members"

shehzan10 replied:
But is this line even really needed?

srothst1 replied:
Ultimately, it is your call on this. Other reviewers seemed fine with this line - it does not make or break the tutorial. For now, I removed the line. Please let me know if you would like me to add it back.

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+- An IDE or code editor. Developers on the Cesium team members use Visual Studio Code, but a minimal code editor such as Sublime Text will also work.
+- Node.js installed. We recommend using the latest LTS version.
+
+## Initialize an app with npm
```
Please use Title case for all headers as much as possible, so in this case:

> Initialize an App with NPM

srothst1 replied:
Just added a commit that addresses this. As noted [above](https://github.com/CesiumGS/cesium-webpack-example/pull/28#discussion_r733716410) npm prefers lower case letters so "Initialize an App with npm" is likely more appropriate there. However, this issue was fixed everywhere else.

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+1. Create a `src` directory for your app code. When we build the app, webpack will automatically produce distribution files in a directory named `dist`.
+2. Create the files `index.html` and `index.js` in the `src` directory added in step 1.
+3. Add the following code to `src/index.html`
+```
````
In Markdown, when adding a code block within a list, add the 4 spaces as a prefix. Makes for better formatting. Also give it a hint for content, in this case html.

srothst1 replied:
I am currently using a [VS Code extension](https://shd101wyy.github.io/markdown-preview-enhanced/#/) to preview my markdown files. I experimented with adding 4 spaces as a prefix before a code block within a list and did not see any changes in the compiled file. I am also not sure what you mean by "Also give it a hint for content, in this case HTML." Any further suggestions?

shehzan10 replied:
I'm pretty sure the VS Code extension handles it correctly, but try it in GitHub web editor + preview as well.

You can give markdown hints about syntax formatting. https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

srothst1 replied:
Yes, my research suggests that the VS Code extension handles the markdown correctly. I am also testing this out on the GitHub web preview. When I preview my markdown files, the formatting looks correct to me.

![image](https://user-images.githubusercontent.com/39531367/138484159-dc639672-f30c-4d64-b932-56299b6fe9dc.png)

It is still unclear to me what edits you are asking me to make here. Providing some more details would be helpful.

srothst1 replied:
Notes from meeting:
- indent images and code blocks
- 4 spaced before each line of a code block to indent 
- add js, css, etc after first ''' in code block

srothst1 replied:
Just added a commit that addresses the topics outlined in our one-on-one meeting :+1:

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+  </body>
+</html>
+```
+This will serve as a boilerplate HTML page to get us started.
````
Put this before the code. Easy to miss.

srothst1 replied:
Done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+```
+console.log('Hello World!');
+```
+This will serve as our boilerplate JavaScript code.
````
Same as above.

srothst1 replied:
Done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+};
+```
+In this code, `context` specifies the base path for your files. `entry` is used to specify bundles and `src/index.js` is our entry point. Webpack will output the bundel `app.js` to the folder `dist`, that webpack will create at runtime.
+
````
Remove empty line between list. Also below.

srothst1 replied:
done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+};
+```
+
+## Bundle the app
````
Add empty line after heading.

srothst1 replied:
done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+}
+```
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
````
You can hyperlink the sections.

srothst1 replied:
done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+```
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
+2. Run the command `npm run build`. Install CLI for webpack if necessary. webpack CLI can be installed using the command `npm install --save-dev webpack-cli`.
````
Why is `webpack` all lowercase at the beginning of the sentence?

srothst1 replied:
Nice catch - just added a commit that addresses this :+1:

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+Please note that details of this json file will vary based on your selections in step 3 of **Initialize an app with npm**.
+2. Run the command `npm run build`. Install CLI for webpack if necessary. webpack CLI can be installed using the command `npm install --save-dev webpack-cli`.
+3. Ensure that there are no errors and your output looks something like
```
`output looks something like` -> `output looks similar to`

srothst1 replied:
I just added a commit that addresses this. It seems like "output looks similar to" is a more professional phrase.

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+./src/index.js 28 bytes [built] [code generated]
+webpack 5.50.0 compiled successfully in 86 ms
+```
+Please note that the `app.js` bundle and `index.html` file should be added to the `dist` folder.
````
Should this be "Please note" or should it be "Please verify that...."

srothst1 replied:
Re-worded this phrase :+1:

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+}
+```
+3. Run `npm start` and open http://localhost:8080/ in a web browser.
+4. Ensure that you see Hello World!
````
Is "Ensure" the right phrasing here?

srothst1 replied:
I changed "Ensure" to "Verify"

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+3. Run `npm start` and open http://localhost:8080/ in a web browser.
+4. Ensure that you see Hello World!
+
+![South San Francisco](./screenshots/hw.png)
```
This file can use a better title and filename.

srothst1 replied:
:+1:

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+## Add CesiumJS to a webpack app
+
+CesiumJS is a large and complex library. In additional to JavaScript modules, it also includes static assets such as CSS, image, and json files. It includes web worker files to perform intensive calculations in separate threads. Unlike traditional npm modules, CesiumJS does not define an entry point because of the diverse ways in which the library is used. We’ll need to configure some additional  ...
```
> CesiumJS is a large and complex library

Not sure this should be the first line. Maybe introduce what Cesium does first, and then follow with this.

srothst1 replied:
I think this is a good idea! I just added a commit that addresses this. I think starting with some of our capabilities makes a lot more sense.

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+First, define where CesiumJS is. This tutorial uses the source code, so webpack can include individual models and trace the dependencies. Alternatively, you can use the built (minified or unminified) version of CesiumJS. However, the modules are already combined and optimized, which gives us less flexibility.
+
+## Install CesiumJS
```
You headers need to form a tree. Right now everything is a level 2 header with `##`

srothst1 replied:
I currently am using level one headers for section overviews. For instance, "Cesium and Webpack" is a level one header. I am using level two headers for subsections such as "Prerequisites" I am not sure exactly what you are asking me to edit here - can you please provide some more information?

shehzan10 replied:
In my opinion, "Install CesiumJS" is a subheading to "Add CesiumJS to a Webpack App". SO it should be level 3.

srothst1 replied:
Done!

shehzan10 commented on TUTORIAL.md:
````diff
@@ -0,0 +1,396 @@
+};
+```
+
+## Incorporate CesiumJS
````
This is very vague

srothst1 replied:
I added a commit that changes this to "Incorporate CesiumJS Into Your Application"

shehzan10 replied:
Incorporate isn't usually used in a coding sense. Are you sure its the right term here?

srothst1 replied:
To be honest, "incorporate" reads fine to me in this context. However, I want to make sure that this demo is perfect for our community - is there another term that you would prefer? 

"Add CesiumJS to your Application"
"Include CesiumJS in your Application"

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+3. Run the command `npm run build`. Ensure that the project is built correctly.
+4. Run `npm start` and open http://localhost:8080/ in a web browser to see the CesiumJS viewer.
+
+![South San Francisco](./screenshots/ssf2.png)
```
For image file names, please be more descriptive.

srothst1 replied:
Done :+1:

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+![South San Francisco](./screenshots/ssf2.png)
+
+Copy and paste your favorite Sandcastle example. For example, [The Particle System Fireworks](https://sandcastle.cesium.com/?src=Particle%20System%20Fireworks.html) demo makes for a fun demo.
```
Whats the reasoning behind recommending this as the starting point? I don't think "fun demo" is right.

srothst1 replied:
The last version of this tutorial also recommended this demo. I personally like the fireworks demo a lot. It is very simple but also very flashy. Occasionally, I use it to get new community members enthusiastic about CesiumJS. Regardless, I added a commit that tweaks the wording here.

shehzan10 replied:
Thats completely the reverse way. My point was that you should be suggesting a better fit sandcastle to the reader, not fixing the wording. Please pick a different sandcastle that highlights Cesium's capabilities.

srothst1 replied:
Is the Map Pins demo more appropriate?
https://sandcastle.cesium.com/?src=Map%20Pins.html&label=Beginner
Webpack helps users set up a robust development environment, so there aren't really any demos that are more or less relevant to this tutorial about CesiumJS and Webpack. However, it is important to showcase Cesium's capabilities. I suspect that a demo with the label "Beginner" is likely best.

srothst1 replied:
Notes from meeting:
- use OSM buildings demo

srothst1 replied:
Just added a commit that updated the demo that we are showcasing :+1:

shehzan10 commented on TUTORIAL.md:
```diff
@@ -0,0 +1,396 @@
+
+Source maps allow webpack to trace errors back to the original content. They offer more or less detailed debugging information in exchange for compiling speed. We recommend setting `devtool` to the 'eval' option in `webpack.config.js`
+
+` devtool: 'eval' `
```
Either use triple tick marks for blocks, or single tick marks for inline code.

srothst1 replied:
Done!

shehzan10 commented on src/index.js:
```diff
@@ -1,18 +1,24 @@
-import "./css/main.css";
+// Your access token can be found at: https://cesium.com/ion/tokens.
+// This is the default access token from your ion account
+Ion.defaultAccessToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiJlYWE1OWUxNy1mMWZiLTQzYjYtYTQ0OS1kMWFjYmFkNjc5YzciLCJpZCI6NTc3MzMsImlhdCI6MTYyNzg0NTE4Mn0.XcKpgANiY19MC4bdFUXMVEBToBmqS8kuYpUlxJHYZxk';
```
Where is this default token coming from? Are you adding the right documentation needed to cycle this?

srothst1 replied:
The default token is coming from the [CesiumJS Quickstart Guide](https://cesium.com/learn/cesiumjs-learn/cesiumjs-quickstart/). I think this is appropriate here, given that I am using the code from that guide. Please let me know if you would like me to make any edits here.

shehzan10 replied:
then the comment `This is the default access token from your ion account` is invalid. Its not a token from the users account.

srothst1 replied:
I just updated the wording for the Webpack tutorial. However, in that case, the wording for the CesiumJS quickstart guide is also incorrect and should be updated.

shehzan10 reviewed (changes requested):
@srothst1 Most of looks good on the code side. I made a bunch of comments on the readme side.

srothst1 committed:
updated URLs in README file

srothst1 committed:
changed wording from we -> you

srothst1 committed:
updated npm link phrasing

srothst1 committed:
added a You'll learn how to section

srothst1 committed:
removed line about code text editor

srothst1 committed:
using Tile case for all headers

srothst1 committed:
moved boilerplate phrase above code

srothst1 committed:
Remove empty line between list. Also below.

srothst1 committed:
Add empty line after heading

srothst1 committed:
hyperlinked sections

srothst1 committed:
made webpack upper case

srothst1 committed:
changed output wording

srothst1 committed:
changed wording around app.js bundle

srothst1 committed:
verify not ensure

srothst1 committed:
updated filename

srothst1 committed:
updated working around

srothst1 committed:
changed wording of Incorporate CesiumJS

srothst1 committed:
more descriptive filenames

srothst1 committed:
updated wording around demo showcase

srothst1 committed:
updated md syntax

srothst1 committed:
updated token wording

srothst1 committed:
install cesiumjs is level three header

srothst1 committed:
added a commit that fixed the markdown syntax and formatting

srothst1 committed:
updated the sandcastle demo that we are showcasing

shehzan10 commented:
@srothst1 Thanks for the changes. @ggetz Feel free to merge if all the tech work is reviewed and fixed.

ggetz commented:
Tech aspect is all ready to go from my view. Thanks @srothst1 for seeing this through!

TJKoury commented:
[C137.js is now open-sourced and Apache-2.0 licensed.](https://github.com/DigitalArsenal/c137.js)

thw0rted commented:
Thanks for the pointer TJ!  I'm looking over it now.  Where's the best place to ask questions / leave comments?  I'd love to see some of the ideas you use incorporated into the way Cesium handles their own packaging...

TJKoury commented:
@thw0rted Start the discussion on the repo itself, open an issue.

maylortaylor commented:
Uncaught Error: Cannot find module `'..\..\node_modules\cesium/Build/CesiumUnminified/index.cjs'`  at the point below:
module.exports = __webpack_require__("../../node_modules/cesium sync recursive")(path.join(
  __dirname,
  "Build/CesiumUnminified/index.cjs"
));

TJKoury commented:
Just FYI, we deprecated c137.js since no one was using it and the changes to the build process obsoleted our approach.  The core Cesium team is aware of these requirements and are working to incorporate some of the concepts I explored into the current official build process.
